core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java

```diff
@@ -48,154 +45,195 @@
-     * @param localCheckpoint the local checkpoint for the shard
-     * @param map the tracking map
-     * @param reason the reason for the update (used for logging)
-     * @return {@code true} if the local checkpoint was updated, otherwise {@code false} if this was a no-op
+     * Class invariant that should hold before and after every invocation of public methods on this class. As Java lacks implication
+     * as a logical operator, many of the invariants are written under the form (!A || B), they should be read as (A implies B) however.
      */
-    private boolean updateLocalCheckpoint(
-            final String allocationId, final long localCheckpoint, ObjectLongMap<String> map, final String reason) {
-        final int index = map.indexOf(allocationId);
-        if (index >= 0) {
-            final long current = map.indexGet(index);
-            if (current < localCheckpoint) {
-                map.indexReplace(index, localCheckpoint);
-                logger.trace("updated local checkpoint of [{}] in [{}] from [{}] to [{}]", allocationId, reason, current, localCheckpoint);
-            } else {
-                logger.trace(
-                        "skipped updating local checkpoint of [{}] in [{}] from [{}] to [{}], current checkpoint is higher",
-                        allocationId,
-                        reason,
-                        current,
-                        localCheckpoint);
-            }
-            return true;
-        } else {
-            return false;
+    private boolean invariant() {
+        /**
+         * local checkpoints only set during primary mode
+         */
+        assert primaryMode || localCheckpoints.values().stream()
+            .allMatch(lcps -> lcps.localCheckPoint == SequenceNumbersService.UNASSIGNED_SEQ_NO ||
+                lcps.localCheckPoint == SequenceNumbersService.PRE_60_NODE_LOCAL_CHECKPOINT);
+        /**
+         * relocation handoff can only occur in primary mode
+         */
+        assert !handOffInProgress || primaryMode;
+        /**
+         * there is at least one in-sync shard copy when the global checkpoint tracker operates in primary mode (i.e. the shard itself)
+         */
+        assert !primaryMode || localCheckpoints.values().stream().anyMatch(lcps -> lcps.inSync);
+        /**
+         * during relocation handoff there are no entries blocking global checkpoint advancement
+         */
+        assert !handOffInProgress || pendingInSync.isEmpty();
+        /**
+         * entries blocking global checkpoint advancement can only exist in primary mode and when not having a relocation handoff
+         */
+        assert pendingInSync.isEmpty() || (primaryMode && !handOffInProgress);
+        /**
+         * the computed global checkpoint is always up-to-date
+         */
+        assert !primaryMode || globalCheckpoint == computeGlobalCheckPoint(pendingInSync, localCheckpoints.values(), globalCheckpoint);
```

Comment:
can we add a message with the globalCheckpoint and the result of the computation?